tensorflow/compiler/xla/stream_executor/stream.cc

```diff
@@ -1427,6 +1429,117 @@ Stream &Stream::ThenBlasSbmv(blas::UpperLower uplo, uint64_t n, uint64 k,
               x, incx, beta, y, incy);
 }
 
+template <typename InputType>
+tsl::Status Stream::ThenBlasGemm(blas::Transpose transa, blas::Transpose transb,
+                          uint64_t m, uint64 n, uint64 k,
+                          const DeviceMemory<InputType> &a, int lda,
+                          const DeviceMemory<InputType> &b, int ldb,
+                          DeviceMemory<InputType> *c, int ldc,
+                          blas::ComputePrecision precision,
+                          blas::CallContext context) {
+  InputType alpha{1.0};
+  InputType beta{0.0};
+  if(gpu::GpuBlasLtEnabled()) {
+    auto& r = gpu::BlasLtGemmRunner::i(this);
+    CheckStatus(r.Run(*this, transa, transb, m, n, k, 
+      alpha, a, lda, b, ldb, beta, c, ldc, 
+      /* allocator */nullptr)); //! NOTE: allocator is not available!!
+    return ::tsl::OkStatus();
+  }
+  return ThenBlasGemm(transa, transb, m, n, k, alpha, a, lda, b, ldb, beta, c,
+                      ldc, precision, context);
+}
+
+#define INSTANTIATE_THEN_BLAS_GEMM(INPUT_TYPE)                                 \
+    template tsl::Status Stream::ThenBlasGemm<INPUT_TYPE>(                     \
+        blas::Transpose transa, blas::Transpose transb,                        \
+        uint64_t m, uint64 n, uint64 k,                                        \
+        const DeviceMemory<INPUT_TYPE>& a, int lda,                            \
+        const DeviceMemory<INPUT_TYPE>& b, int ldb,                            \
+        DeviceMemory<INPUT_TYPE>* c, int ldc,                                  \
+        blas::ComputePrecision precision,                                      \
+        blas::CallContext context);
+
+INSTANTIATE_THEN_BLAS_GEMM(float)
```

Comment:
I like the idea of moving template implementations to the CPP file with explicit instantiation.. Then we have less dependencies in stream.h file.